### PR TITLE
msg/async/rdma: fix bug event center is blocked by rdma construct connection for transport ib sync msg

### PR DESCRIPTION
**Pull request by liupengs**
We construct a tcp connection to transport ib sync msg, if the
remote node is shutdown (shutdown by accident), the net.connect will be blocked until timeout
is reached, which cause the event center be blocked.

This bug may cause mon probe timeout and osd not reply, and so on.

Fixes: https://tracker.ceph.com/issues/42452
Signed-off-by: Peng Liu <liupeng37@baidu.com>

### CODE REVIEW

File: src/msg/async/rdma/RDMAStack.h [hunk 1]
```diff
@@ -235,6 +237,20 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
       active = false;
     }
   };
+  class C_handle_connection_init : public EventCallback {
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]: no need to define this callback in the header file. please define it in an anonymous namespace in `RDMAConnectedSocketImpl.cc`.

File: src/msg/async/rdma/RDMAStack.h [hunk 2]
```diff
@@ -235,6 +237,20 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
...
+    RDMAConnectedSocketImpl *csi;
+    bool active;
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]: nit, please init this variable in-class, like
```c++
bool active = true;
```

File: src/msg/async/rdma/RDMAStack.h [hunk 3]
```diff
@@ -235,6 +237,20 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
...
+   public:
+    C_handle_connection_init(RDMAConnectedSocketImpl *w) : csi(w), active(true) {}
+    void do_request(uint64_t fd) {
+      if (active)
+        csi->handle_connection_init();
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]: why fault the connection?

**Comment by tchaikov** [comment]
@changchengx mind taking a look?

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 4]
```diff
@@ -172,7 +173,14 @@ int RDMAConnectedSocketImpl::try_connect(const entity_addr_t& peer_addr, const S
   ldout(cct, 20) << __func__ << " nonblock:" << opts.nonblock << ", nodelay:"
                  << opts.nodelay << ", rbuf_size: " << opts.rcbuf_size << dendl;
   NetHandler net(cct);
-  tcp_fd = net.connect(peer_addr, opts.connect_bind_addr);
+
+  // we construct a socket to transport ib sync message
+  // but we shouldn't block in tcp connecting
+  if (opts.nonblock) {
```
[the remark this answers is not in the document]
Comment (liupengs) [reply src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: yes, master has the same problem, I will fix the master this weekend and repost this PR。

File: src/msg/async/rdma/RDMAStack.h (hunk 3 above)
Comment (liupengs) [reply src/msg/async/rdma/RDMAStack.h]: if this conneciton is connected,  the function handle_connection_init should't be called,  except some logic error , such as the con_init_handler is not deleted from the event center. 
if we failed to send_msg,  I think we need to fault the connection.

File: src/msg/async/rdma/RDMAStack.h (hunk 1 above)
Comment (liupengs) [reply src/msg/async/rdma/RDMAStack.h]: don't need keep C_handle_connection_init in the  same format as C_handle_connection?

**Comment by changchengx** [comment]
@liupengs Please verify this patch does't break "ms_async_rdma_type=iwarp".
You could use do below configuration when trying to check it:
1) ms_type = async+rdma
2) ms_async_rdma_type=iwarp
3) ms_async_rdma_cm=true

Other info (not related with patch)
1. The "RDMAIWARP*" class name is misleading since the real implementation does not only support iWARP but also supports RoCEv2.
2. The main target of "RDMAIWARP*" class is to use RDAMCM library to manage the connection because that iWARP connection is only supported by RDAMCM library. However, RDMACM library also support  RoCEv2 connection.
3. Technically speaking, it's not good to use keep using RDMAConnectedSocketImpl/RDMAServerSocketImpl since they use ib_cm_meta_t to exchange the meta data while these info should be exchanged by RDMACM library. Glusterfs also use RDMACM to manage the connection.

**Comment by tchaikov** [comment]
@liupengs ping?

**Comment by liupengs** [comment]
Sorry, I am very busy recently, I will complete the test within one week.

**Comment by liupengs** [comment]
> @liupengs Please verify this patch does't break "ms_async_rdma_type=iwarp".
> You could use do below configuration when trying to check it:
> 
> 1. ms_type = async+rdma
> 2. ms_async_rdma_type=iwarp
> 3. ms_async_rdma_cm=true
> 
> Other info (not related with patch)
> 
> 1. The "RDMAIWARP*" class name is misleading since the real implementation does not only support iWARP but also supports RoCEv2.
> 2. The main target of "RDMAIWARP*" class is to use RDAMCM library to manage the connection because that iWARP connection is only supported by RDAMCM library. However, RDMACM library also support  RoCEv2 connection.
> 3. Technically speaking, it's not good to use keep using RDMAConnectedSocketImpl/RDMAServerSocketImpl since they use ib_cm_meta_t to exchange the meta data while these info should be exchanged by RDMACM library. Glusterfs also use RDMACM to manage the connection.

I tested this branch with configure :

**Comment by tchaikov** [comment]
@liupengs thanks for testing this change.

**Comment by liupengs** [comment]
@tchaikov Please review this code.

File: src/msg/async/rdma/RDMAStack.h [hunk 5]
```diff
@@ -246,6 +248,20 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
       active = false;
     }
   };
+  class C_handle_connection_init : public EventCallback {
+    RDMAConnectedSocketImpl *csi;
+    bool active = true;
+   public:
+    C_handle_connection_init(RDMAConnectedSocketImpl *w) : csi(w) {}
+    void do_request(uint64_t fd) {
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]:
```suggestion
    void do_request(uint64_t fd) final {
```

File: src/msg/async/rdma/RDMAStack.h [hunk 6]
```diff
@@ -189,6 +189,7 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
   std::vector<ibv_wc> wc;
   bool is_server;
   EventCallbackRef con_handler;
+  EventCallbackRef con_init_handler;
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]: why not use a `unique_ptr<EventCallback>`?

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 7]
```diff
@@ -517,11 +551,16 @@ void RDMAConnectedSocketImpl::cleanup() {
...
-      worker->center.delete_file_event(tcp_fd, EVENT_READABLE);
+      worker->center.delete_file_event(tcp_fd, EVENT_READABLE | EVENT_WRITABLE);
     }, false);
     delete con_handler;
     con_handler = nullptr;
   }
+  if (con_init_handler) {
+    (static_cast<C_handle_connection_init*>(con_init_handler))->close();
+    delete con_init_handler;
+    con_init_handler = nullptr;
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: no need to check for null before `delete con_init_handler`. why would you need to reset it as the only caller of this function is `~RDMAConnectedSocketImpl()`. so nobody will get a chance to dereference it after that.

Comment (tchaikov) [reply src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: for the same reason, no need to `close()` `con_init_handler` for resetting "active".

File: src/msg/async/rdma/RDMAStack.h (hunk 6 above)
Comment (tchaikov) [reply src/msg/async/rdma/RDMAStack.h]: s/con_init_handler/established_handler/

and add a commit to s/con_handler/read_handler/

File: src/msg/async/rdma/RDMAStack.h [hunk 8]
```diff
@@ -235,9 +237,9 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
 
   class C_handle_connection : public EventCallback {
     RDMAConnectedSocketImpl *csi;
-    bool active;
+    bool active = true;
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]: please move this change to a separated commit.

File: src/msg/async/rdma/RDMAStack.h (hunk 1 above)
Comment (tchaikov) [reply src/msg/async/rdma/RDMAStack.h]: good idea! better off moving `C_handle_connection` to the .cc file as well, please do this in a separated commit.

File: src/msg/async/rdma/RDMAStack.h (hunk 3 above)
Comment (tchaikov) [reply src/msg/async/rdma/RDMAStack.h]: i think we'd better add an assert here like
```c++
if (active) {
  assert(!csi->is_connected());
  csi->handle_connection_established();
```

File: src/msg/async/rdma/RDMAStack.h [hunk 9]
```diff
@@ -225,6 +226,7 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
   int activate();
   void fin();
   void handle_connection();
+  int handle_connection_init(bool need_set_fault = true);
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]: s/handle_connection_init/handle_connection_established/

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc (hunk 7 above)
Comment (liupengs) [reply src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: > no need to check for null before `delete con_init_handler`. why would you need to reset it as the only caller of this function is `~RDMAConnectedSocketImpl()`. so nobody will get a chance to dereference it after that.

`cleanup()`  maybe be called by ` RDMAConnectedSocketImpl::read()`。

File: src/msg/async/rdma/RDMAStack.h (hunk 6 above)
Comment (liupengs) [reply src/msg/async/rdma/RDMAStack.h]: > why not use a `unique_ptr<EventCallback>`?


> why not use a `unique_ptr<EventCallback>`?
1.  keep it the style  with the old code; 
2.  we don't need to use a unique_ptr<EventCallback>,  I will free con_init_handler in `cleanup()`.

File: src/msg/async/rdma/RDMAStack.h (hunk 3 above)
Comment (liupengs) [reply src/msg/async/rdma/RDMAStack.h]: > i think we'd better add an assert here like
> 
> ```c++
> if (active) {
>   assert(!csi->is_connected());
>   csi->handle_connection_established();
> ```

we should not call `assert(!csi->is_connected())` in here,  `handle_connection_established()` will check the connection status.

**Pull request merged**